File: src/main/java/com/tkr/wire/SessionWireCodec.java
package com.tkr.wire;

import com.tkr.types.WireTypes;
import com.tkr.types.WireTypes.*;
import com.tkr.util.BoundsUtil;
import com.tkr.util.DigestUtil;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/** TKR3 session wire encode/decode per FORMAT.md. */
public final class SessionWireCodec {

    public static final int SESSION_HEADER_SIZE = 24;
    public static final int SESSION_LEG_SIZE = 28;

    public static final class SessionCodecConfig {
        public boolean verifyMergeDigest = true;
        public int maxLegs = WireTypes.MAX_SESSION_LEGS;
    }

    public static final class SessionDecodeResult {
        public Status status = Status.OK;
        public SessionWireFrame frame = new SessionWireFrame();
        public int bytesConsumed;
    }

    public static final class SessionEncodeResult {
        public Status status = Status.OK;
        public byte[] data = new byte[0];
    }

    private final SessionCodecConfig config;
    private int nextSlotId = 1;

    public SessionWireCodec() { this(new SessionCodecConfig()); }
    public SessionWireCodec(SessionCodecConfig config) {
        this.config = config != null ? config : new SessionCodecConfig();
    }

    public SessionDecodeResult decodeSession(byte[] data, int offset, int length) {
        SessionDecodeResult result = new SessionDecodeResult();
        if (data == null || length < SESSION_HEADER_SIZE) {
            result.status = Status.TRUNCATED;
            return result;
        }
        WireSessionHeader header = new WireSessionHeader();
        readHeader(data, offset, header);
        Status vh = validateHeader(header);
        if (vh != Status.OK) { result.status = vh; return result; }
        result.frame.header = header;
        int legBase = offset + SESSION_HEADER_SIZE;
        List<WireSessionLeg> legs = new ArrayList<>();
        Status lr = readLegs(data, offset + length, legBase, header.legCount, legs);
        if (lr != Status.OK) { result.status = lr; return result; }
        result.frame.legs = legs;
        int refStart = legBase + header.legCount * SESSION_LEG_SIZE;
        byte[] refBlob = Arrays.copyOfRange(data, refStart, offset + length);
        result.frame.refBlob = refBlob;
        for (WireSessionLeg leg : legs) {
            Status rb = validateLegBounds(leg, refBlob.length);
            if (rb != Status.OK) { result.status = rb; return result; }
        }
        if (config.verifyMergeDigest && header.mergeDigest != 0) {
            int computed = computeRefDigest(refBlob);
            if (computed != header.mergeDigest) {
                result.status = Status.BOUNDS_ERROR;
                return result;
            }
        }
        buildMergeSlots(result.frame);
        result.bytesConsumed = length;
        return result;
    }

    public SessionDecodeResult decodeSession(byte[] data) {
        return decodeSession(data, 0, data != null ? data.length : 0);
    }

    public SessionEncodeResult encodeSession(SessionWireFrame frame) {
        SessionEncodeResult result = new SessionEncodeResult();
        if (frame == null || frame.legs.size() > config.maxLegs) {
            result.status = Status.BOUNDS_ERROR;
            return result;
        }
        if (frame.refBlob == null) frame.refBlob = new byte[0];
        frame.header.magic = WireTypes.SESSION_MAGIC;
        frame.header.version = WireTypes.WIRE_VERSION;
        frame.header.legCount = frame.legs.size();
        frame.header.mergeDigest = computeRefDigest(frame.refBlob);
        int total = SESSION_HEADER_SIZE + frame.legs.size() * SESSION_LEG_SIZE + frame.refBlob.length;
        byte[] out = new byte[total];
        writeHeader(out, 0, frame.header);
        int cursor = SESSION_HEADER_SIZE;
        for (WireSessionLeg leg : frame.legs) {
            writeLeg(out, cursor, leg);
            cursor += SESSION_LEG_SIZE;
        }
        System.arraycopy(frame.refBlob, 0, out, cursor, frame.refBlob.length);
        result.data = out;
        return result;
    }

    private void readHeader(byte[] data, int offset, WireSessionHeader out) {
        out.magic = BoundsUtil.readU32Le(data, offset);
        out.version = BoundsUtil.readU16Le(data, offset + 4);
        out.legCount = BoundsUtil.readU16Le(data, offset + 6);
        out.sessionId = BoundsUtil.readU32Le(data, offset + 8);
        out.checkpointSeq = BoundsUtil.readU32Le(data, offset + 12);
        out.flags = BoundsUtil.readU32Le(data, offset + 16);
        out.mergeDigest = BoundsUtil.readU32Le(data, offset + 20);
    }

    private void writeHeader(byte[] dst, int offset, WireSessionHeader h) {
        BoundsUtil.writeU32Le(dst, offset, h.magic);
        BoundsUtil.writeU16Le(dst, offset + 4, h.version);
        BoundsUtil.writeU16Le(dst, offset + 6, h.legCount);
        BoundsUtil.writeU32Le(dst, offset + 8, h.sessionId);
        BoundsUtil.writeU32Le(dst, offset + 12, h.checkpointSeq);
        BoundsUtil.writeU32Le(dst, offset + 16, h.flags);
        BoundsUtil.writeU32Le(dst, offset + 20, h.mergeDigest);
    }

    private Status readLegs(byte[] data, int totalSize, int offset, int count, List<WireSessionLeg> out) {
        if (!BoundsUtil.sectionBodyInBounds(offset, (long) count * SESSION_LEG_SIZE, totalSize)) {
            return Status.BOUNDS_ERROR;
        }
        Set<Integer> seen = new HashSet<>();
        for (int i = 0; i < count; i++) {
            int base = offset + i * SESSION_LEG_SIZE;
            WireSessionLeg leg = new WireSessionLeg();
            leg.legId = BoundsUtil.readU32Le(data, base);
            leg.clOrdId = BoundsUtil.readU32Le(data, base + 4);
            leg.allocAccount = BoundsUtil.readU32Le(data, base + 8);
            leg.qtyMilli = BoundsUtil.readU32Le(data, base + 12);
            leg.refOffset = BoundsUtil.readU32Le(data, base + 16);
            leg.refLen = BoundsUtil.readU32Le(data, base + 20);
            leg.flags = BoundsUtil.readU32Le(data, base + 24);
            if (!seen.add(leg.legId)) return Status.DUPLICATE_KEY;
            out.add(leg);
        }
        return Status.OK;
    }

    private void writeLeg(byte[] dst, int offset, WireSessionLeg leg) {
        BoundsUtil.writeU32Le(dst, offset, leg.legId);
        BoundsUtil.writeU32Le(dst, offset + 4, leg.clOrdId);
        BoundsUtil.writeU32Le(dst, offset + 8, leg.allocAccount);
        BoundsUtil.writeU32Le(dst, offset + 12, leg.qtyMilli);
        BoundsUtil.writeU32Le(dst, offset + 16, leg.refOffset);
        BoundsUtil.writeU32Le(dst, offset + 20, leg.refLen);
        BoundsUtil.writeU32Le(dst, offset + 24, leg.flags);
    }

    private Status validateHeader(WireSessionHeader header) {
        if (header.magic != WireTypes.SESSION_MAGIC) return Status.INVALID_MAGIC;
        if (header.version != WireTypes.WIRE_VERSION) return Status.BOUNDS_ERROR;
        if (header.legCount > config.maxLegs) return Status.BOUNDS_ERROR;
        return Status.OK;
    }

    private Status validateLegBounds(WireSessionLeg leg, int refSize) {
        if (leg.refLen == 0) return Status.OK;
        return BoundsUtil.sliceInBounds(leg.refOffset, leg.refLen, refSize) ? Status.OK : Status.BOUNDS_ERROR;
    }

    private void buildMergeSlots(SessionWireFrame frame) {
        frame.mergeSlots.clear();
        for (WireSessionLeg leg : frame.legs) {
            if ((leg.flags & WireTypes.LEG_FLAG_MERGE_PENDING) == 0 || leg.refLen == 0) continue;
            MergeSlot slot = new MergeSlot();
            slot.slotId = nextSlotId++;
            slot.legId = leg.legId;
            slot.refLen = leg.refLen;
            slot.refData = Arrays.copyOfRange(frame.refBlob, leg.refOffset, leg.refOffset + leg.refLen);
            slot.checkpointSeq = frame.header.checkpointSeq;
            slot.pinned = false;
            frame.mergeSlots.add(slot);
        }
    }

    public int computeRefDigest(byte[] refBlob) {
        return DigestUtil.fnv1a32(refBlob);
    }

    public int computeLegTableDigest(List<WireSessionLeg> legs) {
        int hash = DigestUtil.FNV_OFFSET;
        for (WireSessionLeg leg : legs) {
            hash = DigestUtil.mixDigest(hash, leg.legId);
            hash = DigestUtil.mixDigest(hash, leg.allocAccount);
            hash = DigestUtil.mixDigest(hash, leg.qtyMilli);
        }
        return hash;
    }

    public long sumQtyMilli(SessionWireFrame frame) {
        long sum = 0;
        for (WireSessionLeg leg : frame.legs) sum += leg.qtyMilli;
        return sum;
    }

    public int computeHeaderDigest(WireSessionHeader header) {
        byte[] buf = new byte[SESSION_HEADER_SIZE];
        BoundsUtil.writeU32Le(buf, 0, header.magic);
        BoundsUtil.writeU16Le(buf, 4, header.version);
        BoundsUtil.writeU16Le(buf, 6, header.legCount);
        BoundsUtil.writeU32Le(buf, 8, header.sessionId);
        BoundsUtil.writeU32Le(buf, 12, header.checkpointSeq);
        BoundsUtil.writeU32Le(buf, 16, header.flags);
        BoundsUtil.writeU32Le(buf, 20, 0);
        return DigestUtil.fnv1a32(buf, 0, SESSION_HEADER_SIZE);
    }

    public Status stageMergeSlots(SessionWireFrame frame) {
        if (frame == null) return Status.BOUNDS_ERROR;
        frame.mergeSlots.clear();
        nextSlotId = 1;
        for (WireSessionLeg leg : frame.legs) {
            if (leg.refLen == 0) continue;
            if (!BoundsUtil.sliceInBounds(leg.refOffset, leg.refLen, frame.refBlob.length)) return Status.BOUNDS_ERROR;
            MergeSlot slot = new MergeSlot();
            slot.slotId = nextSlotId++;
            slot.legId = leg.legId;
            slot.refLen = leg.refLen;
            slot.refData = Arrays.copyOfRange(frame.refBlob, leg.refOffset, leg.refOffset + leg.refLen);
            slot.checkpointSeq = frame.header.checkpointSeq;
            slot.pinned = (leg.flags & WireTypes.LEG_FLAG_CHECKPOINT_PIN) != 0;
            frame.mergeSlots.add(slot);
            if (frame.mergeSlots.size() > WireTypes.MAX_DEFERRED_SLOTS) return Status.BOUNDS_ERROR;
        }
        return Status.OK;
    }

    public Status clearMergeSlots(SessionWireFrame frame) {
        if (frame == null) return Status.BOUNDS_ERROR;
        for (MergeSlot slot : frame.mergeSlots) {
            slot.refData = null;
            slot.pinned = false;
        }
        frame.mergeSlots.clear();
        return Status.OK;
    }

    public SessionEncodeResult encodeLegsOnly(List<WireSessionLeg> legs, byte[] refBlob) {
        WireSessionHeader header = new WireSessionHeader();
        header.magic = WireTypes.SESSION_MAGIC;
        header.version = WireTypes.WIRE_VERSION;
        header.legCount = legs.size();
        header.sessionId = 1;
        header.checkpointSeq = 1;
        header.flags = WireTypes.LEG_FLAG_MERGE_PENDING;
        SessionWireFrame frame = new SessionWireFrame();
        frame.header = header;
        frame.legs = legs;
        frame.refBlob = refBlob != null ? refBlob : new byte[0];
        return encodeSession(frame);
    }

    public Status patchLegRefOffset(WireSessionLeg leg, int newOffset) {
        if (leg == null) return Status.BOUNDS_ERROR;
        leg.refOffset = newOffset;
        return Status.OK;
    }

    public Status relocateRefBlob(SessionWireFrame frame) {
        if (frame == null) return Status.BOUNDS_ERROR;
        byte[] compact = new byte[0];
        java.io.ByteArrayOutputStream out = new java.io.ByteArrayOutputStream();
        for (WireSessionLeg leg : frame.legs) {
            if (leg.refLen == 0) {
                leg.refOffset = out.size();
                continue;
            }
            if (!BoundsUtil.sliceInBounds(leg.refOffset, leg.refLen, frame.refBlob.length)) return Status.BOUNDS_ERROR;
            int oldOffset = leg.refOffset;
            leg.refOffset = out.size();
            out.writeBytes(Arrays.copyOfRange(frame.refBlob, oldOffset, oldOffset + leg.refLen));
        }
        frame.refBlob = out.toByteArray();
        return Status.OK;
    }

    public SessionDecodeResult decodePartial(byte[] data, int offset, int length) {
        SessionDecodeResult result = new SessionDecodeResult();
        if (length < SESSION_HEADER_SIZE + SESSION_LEG_SIZE) {
            result.status = Status.PARTIAL_FRAME;
            return result;
        }
        return decodeSession(data, offset, length);
    }

    public static int estimateEncodedSize(int legCount, int refSize) {
        return SESSION_HEADER_SIZE + legCount * SESSION_LEG_SIZE + refSize;
    }

    public Status validateLegRefs(SessionWireFrame frame) {
        for (WireSessionLeg leg : frame.legs) {
            Status st = validateLegBounds(leg, frame.refBlob.length);
            if (st != Status.OK) return st;
        }
        return Status.OK;
    }

    private void appendLegBytes(java.io.ByteArrayOutputStream out, WireSessionLeg leg) throws java.io.IOException {
        byte[] buf = new byte[4];
        BoundsUtil.writeU32Le(buf, 0, leg.legId);
        out.write(buf);
        BoundsUtil.writeU32Le(buf, 0, leg.clOrdId);
        out.write(buf);
        BoundsUtil.writeU32Le(buf, 0, leg.allocAccount);
        out.write(buf);
        BoundsUtil.writeU32Le(buf, 0, leg.qtyMilli);
        out.write(buf);
        BoundsUtil.writeU32Le(buf, 0, leg.refOffset);
        out.write(buf);
        BoundsUtil.writeU32Le(buf, 0, leg.refLen);
        out.write(buf);
        BoundsUtil.writeU32Le(buf, 0, leg.flags);
        out.write(buf);
    }
// --- expanded helpers ---
    public SessionWireFrame createMinimalFrame(int sessionId, int legCount) {
        SessionWireFrame frame = new SessionWireFrame();
        frame.header.sessionId = sessionId;
        for (int i = 0; i < legCount; i++) {
            WireSessionLeg leg = new WireSessionLeg();
            leg.legId = i + 1;
            leg.qtyMilli = 1000;
            frame.legs.add(leg);
        }
        return frame;
    }
}

File: legacy-cpp/src/wire/session_wire_codec.cc
#include "tkr/wire/session_wire_codec.h"

#include "tkr/util/bounds.h"

#include <algorithm>
#include <cstring>

namespace tkr {
namespace wire {
namespace {

constexpr std::size_t kSessionHeaderSize = 24;
constexpr std::size_t kSessionLegSize = 28;

std::uint32_t ReadU32Le(const std::uint8_t* data) {
  return static_cast<std::uint32_t>(data[0]) |
         (static_cast<std::uint32_t>(data[1]) << 8) |
         (static_cast<std::uint32_t>(data[2]) << 16) |
         (static_cast<std::uint32_t>(data[3]) << 24);
}

std::uint16_t ReadU16Le(const std::uint8_t* data) {
  return static_cast<std::uint16_t>(data[0]) |
         (static_cast<std::uint16_t>(data[1]) << 8);
}

void WriteU32Le(std::uint8_t* dst, std::uint32_t value) {
  dst[0] = static_cast<std::uint8_t>(value & 0xFFu);
  dst[1] = static_cast<std::uint8_t>((value >> 8) & 0xFFu);
  dst[2] = static_cast<std::uint8_t>((value >> 16) & 0xFFu);
  dst[3] = static_cast<std::uint8_t>((value >> 24) & 0xFFu);
}

void WriteU16Le(std::uint8_t* dst, std::uint16_t value) {
  dst[0] = static_cast<std::uint8_t>(value & 0xFFu);
  dst[1] = static_cast<std::uint8_t>((value >> 8) & 0xFFu);
}

}  // namespace

SessionWireCodec::SessionWireCodec(SessionCodecConfig config)
    : config_(config), next_slot_id_(1) {}

void SessionWireCodec::AppendU32Le(std::vector<std::uint8_t>* out,
                                 std::uint32_t value) {
  std::uint8_t buf[4];
  WriteU32Le(buf, value);
  out->insert(out->end(), buf, buf + 4);
}

void SessionWireCodec::AppendU16Le(std::vector<std::uint8_t>* out,
                                   std::uint16_t value) {
  std::uint8_t buf[2];
  WriteU16Le(buf, value);
  out->insert(out->end(), buf, buf + 2);
}

void SessionWireCodec::AppendLegBytes(std::vector<std::uint8_t>* out,
                                      const WireSessionLeg& leg) {
  AppendU32Le(out, leg.leg_id);
  AppendU32Le(out, leg.cl_ord_id);
  AppendU32Le(out, leg.alloc_account);
  AppendU32Le(out, leg.qty_milli);
  AppendU32Le(out, leg.ref_offset);
  AppendU32Le(out, leg.ref_len);
  AppendU32Le(out, leg.flags);
}

std::uint32_t SessionWireCodec::ComputeHeaderDigest(
    const WireSessionHeader& header) const {
  std::uint8_t buf[kSessionHeaderSize];
  WriteU32Le(buf + 0, header.magic);
  WriteU16Le(buf + 4, header.version);
  WriteU16Le(buf + 6, header.leg_count);
  WriteU32Le(buf + 8, header.session_id);
  WriteU32Le(buf + 12, header.checkpoint_seq);
  WriteU32Le(buf + 16, header.flags);
  WriteU32Le(buf + 20, 0);
  return util::Fnv1a32(buf, kSessionHeaderSize);
}

std::uint32_t SessionWireCodec::ComputeRefDigest(const std::uint8_t* data,
                                                 std::size_t len) const {
  if (data == nullptr || len == 0) {
    return util::Fnv1a32(nullptr, 0);
  }
  return util::Fnv1a32(data, len);
}

std::uint32_t SessionWireCodec::ComputeLegTableDigest(
    const std::vector<WireSessionLeg>& legs) const {
  std::uint32_t digest = 2166136261u;
  for (const WireSessionLeg& leg : legs) {
    digest ^= leg.leg_id;
    digest *= 16777619u;
    digest ^= leg.alloc_account;
    digest *= 16777619u;
    digest ^= leg.qty_milli;
    digest *= 16777619u;
  }
  return digest;
}

Status SessionWireCodec::ValidateHeader(const WireSessionHeader& header) const {
  if (header.magic != kSessionMagic) {
    return Status::kInvalidMagic;
  }
  if (header.version != kWireVersion) {
    return Status::kBoundsError;
  }
  if (header.leg_count > config_.max_legs ||
      header.leg_count > kMaxSessionLegs) {
    return Status::kBoundsError;
  }
  return Status::kOk;
}

Status SessionWireCodec::ValidateLegBounds(const WireSessionLeg& leg,
                                           std::size_t ref_size) const {
  if (leg.ref_len == 0) {
    return Status::kOk;
  }
  if (!util::SliceInBounds(leg.ref_offset, leg.ref_len, ref_size)) {
    return Status::kBoundsError;
  }
  return Status::kOk;
}

Status SessionWireCodec::ValidateLegRefs(const SessionWireFrame& frame) const {
  for (const WireSessionLeg& leg : frame.legs) {
    Status st = ValidateLegBounds(leg, frame.ref_blob.size());
    if (st != Status::kOk) {
      return st;
    }
  }
  return Status::kOk;
}

Status SessionWireCodec::ReadHeader(const std::uint8_t* data, std::size_t size,
                                    WireSessionHeader* out,
                                    std::size_t* consumed) {
  if (out == nullptr || data == nullptr || size < kSessionHeaderSize) {
    return Status::kTruncated;
  }

  out->magic = ReadU32Le(data + 0);
  out->version = ReadU16Le(data + 4);
  out->leg_count = ReadU16Le(data + 6);
  out->session_id = ReadU32Le(data + 8);
  out->checkpoint_seq = ReadU32Le(data + 12);
  out->flags = ReadU32Le(data + 16);
  out->merge_digest = ReadU32Le(data + 20);

  if (consumed != nullptr) {
    *consumed = kSessionHeaderSize;
  }
  return Status::kOk;
}

Status SessionWireCodec::ReadLegs(const std::uint8_t* data, std::size_t size,
                                  std::size_t offset, std::uint16_t count,
                                  std::vector<WireSessionLeg>* out) {
  if (out == nullptr || data == nullptr) {
    return Status::kBoundsError;
  }

  const std::size_t table_bytes = static_cast<std::size_t>(count) * kSessionLegSize;
  if (!util::SectionBodyInBounds(offset, table_bytes, size)) {
    return Status::kBoundsError;
  }

  out->clear();
  out->reserve(count);
  std::size_t cursor = offset;

  for (std::uint16_t i = 0; i < count; ++i) {
    WireSessionLeg leg{};
    leg.leg_id = ReadU32Le(data + cursor + 0);
    leg.cl_ord_id = ReadU32Le(data + cursor + 4);
    leg.alloc_account = ReadU32Le(data + cursor + 8);
    leg.qty_milli = ReadU32Le(data + cursor + 12);
    leg.ref_offset = ReadU32Le(data + cursor + 16);
    leg.ref_len = ReadU32Le(data + cursor + 20);
    leg.flags = ReadU32Le(data + cursor + 24);
    out->push_back(leg);
    cursor += kSessionLegSize;
  }

  return Status::kOk;
}

Status SessionWireCodec::ReadRefBlob(const std::uint8_t* data, std::size_t size,
                                     std::size_t offset,
                                     std::vector<std::uint8_t>* out) {
  if (out == nullptr) {
    return Status::kBoundsError;
  }
  if (offset >= size) {
    out->clear();
    return Status::kOk;
  }
  out->assign(data + offset, data + size);
  return Status::kOk;
}

SessionEncodeResult SessionWireCodec::EncodeSession(
    const WireSessionHeader& header, const std::vector<WireSessionLeg>& legs,
    const std::uint8_t* ref_data, std::size_t ref_size) {
  SessionEncodeResult result{};
  result.status = ValidateHeader(header);
  if (result.status != Status::kOk) {
    return result;
  }

  if (legs.size() != header.leg_count) {
    result.status = Status::kBoundsError;
    return result;
  }

  std::vector<std::uint8_t> wire;
  wire.reserve(kSessionHeaderSize + legs.size() * kSessionLegSize + ref_size);

  WireSessionHeader hdr = header;
  hdr.merge_digest = ComputeLegTableDigest(legs);
  if (ref_data != nullptr && ref_size > 0) {
    hdr.merge_digest ^= ComputeRefDigest(ref_data, ref_size);
  }

  AppendU32Le(&wire, hdr.magic);
  AppendU16Le(&wire, hdr.version);
  AppendU16Le(&wire, hdr.leg_count);
  AppendU32Le(&wire, hdr.session_id);
  AppendU32Le(&wire, hdr.checkpoint_seq);
  AppendU32Le(&wire, hdr.flags);
  AppendU32Le(&wire, hdr.merge_digest);

  for (const WireSessionLeg& leg : legs) {
    AppendLegBytes(&wire, leg);
  }

  if (ref_data != nullptr && ref_size > 0) {
    wire.insert(wire.end(), ref_data, ref_data + ref_size);
  }

  result.wire_bytes = std::move(wire);
  result.merge_digest = hdr.merge_digest;
  result.ref_blob_size = ref_size;
  result.status = Status::kOk;
  return result;
}

SessionDecodeResult SessionWireCodec::DecodeSession(const std::uint8_t* data,
                                                    std::size_t size) {
  SessionDecodeResult result{};
  std::size_t consumed = 0;

  result.status = ReadHeader(data, size, &result.frame.header, &consumed);
  if (result.status != Status::kOk) {
    return result;
  }

  result.status = ValidateHeader(result.frame.header);
  if (result.status != Status::kOk) {
    return result;
  }

  result.status = ReadLegs(data, size, consumed, result.frame.header.leg_count,
                             &result.frame.legs);
  if (result.status != Status::kOk) {
    return result;
  }

  consumed += static_cast<std::size_t>(result.frame.header.leg_count) *
              kSessionLegSize;

  result.status = ReadRefBlob(data, size, consumed, &result.frame.ref_blob);
  if (result.status != Status::kOk) {
    return result;
  }

  if (config_.validate_ref_bounds) {
    result.status = ValidateLegRefs(result.frame);
    if (result.status != Status::kOk) {
      return result;
    }
  }

  if (config_.enable_merge_staging) {
    result.status = StageMergeSlots(&result.frame);
    if (result.status != Status::kOk) {
      return result;
    }
  }

  result.consumed_bytes = size;
  return result;
}

Status SessionWireCodec::StageMergeSlots(SessionWireFrame* frame) {
  if (frame == nullptr) {
    return Status::kBoundsError;
  }

  frame->merge_slots.clear();
  merge_table_.clear();
  next_slot_id_ = 1;

  for (const WireSessionLeg& leg : frame->legs) {
    if (leg.ref_len == 0) {
      continue;
    }
    if (!util::SliceInBounds(leg.ref_offset, leg.ref_len,
                             frame->ref_blob.size())) {
      return Status::kBoundsError;
    }

    MergeSlot slot{};
    slot.slot_id = next_slot_id_++;
    slot.leg_id = leg.leg_id;
    slot.ref_ptr = frame->ref_blob.data() + leg.ref_offset;
    slot.ref_len = leg.ref_len;
    slot.checkpoint_seq = frame->header.checkpoint_seq;
    slot.pinned = (leg.flags & kLegFlagCheckpointPin) != 0;

    frame->merge_slots.push_back(slot);
    merge_table_.push_back(slot);

    if (merge_table_.size() > kMaxDeferredSlots) {
      return Status::kBoundsError;
    }
  }

  return Status::kOk;
}

Status SessionWireCodec::ClearMergeSlots(SessionWireFrame* frame) {
  if (frame == nullptr) {
    return Status::kBoundsError;
  }
  for (MergeSlot& slot : frame->merge_slots) {
    slot.ref_ptr = nullptr;
    slot.pinned = false;
  }
  for (MergeSlot& slot : merge_table_) {
    slot.ref_ptr = nullptr;
    slot.pinned = false;
  }
  frame->merge_slots.clear();
  merge_table_.clear();
  return Status::kOk;
}

SessionEncodeResult SessionWireCodec::EncodeLegsOnly(
    const std::vector<WireSessionLeg>& legs,
    const std::vector<std::uint8_t>& ref_blob) {
  WireSessionHeader header{};
  header.magic = kSessionMagic;
  header.version = kWireVersion;
  header.leg_count = static_cast<std::uint16_t>(legs.size());
  header.session_id = 1;
  header.checkpoint_seq = 1;
  header.flags = kLegFlagMergePending;
  return EncodeSession(header, legs, ref_blob.data(), ref_blob.size());
}

Status SessionWireCodec::PatchLegRefOffset(WireSessionLeg* leg,
                                           std::uint32_t new_offset) {
  if (leg == nullptr) {
    return Status::kBoundsError;
  }
  leg->ref_offset = new_offset;
  return Status::kOk;
}

Status SessionWireCodec::RelocateRefBlob(SessionWireFrame* frame) {
  if (frame == nullptr) {
    return Status::kBoundsError;
  }

  std::vector<std::uint8_t> compact;
  compact.reserve(frame->ref_blob.size());

  for (WireSessionLeg& leg : frame->legs) {
    if (leg.ref_len == 0) {
      leg.ref_offset = static_cast<std::uint32_t>(compact.size());
      continue;
    }
    if (!util::SliceInBounds(leg.ref_offset, leg.ref_len,
                             frame->ref_blob.size())) {
      return Status::kBoundsError;
    }
    const std::uint8_t* src = frame->ref_blob.data() + leg.ref_offset;
    leg.ref_offset = static_cast<std::uint32_t>(compact.size());
    compact.insert(compact.end(), src, src + leg.ref_len);
  }

  frame->ref_blob.swap(compact);
  return Status::kOk;
}

SessionDecodeResult SessionWireCodec::DecodePartial(const std::uint8_t* data,
                                                    std::size_t size) {
  SessionDecodeResult result{};
  if (size < kSessionHeaderSize + kSessionLegSize) {
    result.status = Status::kPartialFrame;
    return result;
  }
  return DecodeSession(data, size);
}

std::uint32_t SessionWireCodec::EstimateEncodedSize(
    std::uint16_t leg_count, std::size_t ref_size) {
  return static_cast<std::uint32_t>(kSessionHeaderSize +
                                    static_cast<std::size_t>(leg_count) *
                                        kSessionLegSize +
                                    ref_size);
}

}  // namespace wire
}  // namespace tkr
